daemon/graphdriver/overlay/overlay.go

```diff
@@ -420,7 +467,13 @@ func (d *Driver) ApplyDiff(id string, parent string, diff archive.Reader) (size
 		}
 	}()
 
-	if err = copyDir(parentRootDir, tmpRootDir, copyHardlink); err != nil {
+	// if quota is used, hardlinks between different volumes are not supported.
```

Comment:
I think we should limit this to just the upper layer as is the case with overlay2. We can set the quota on the `upper` directory and not worry about the potential for project to apply to a root directory. Any directory that has `ApplyDiff` called on it will not be used as a container layer and the `root` directory will never be writable from a container.